### ISSUE TITLE
Crash opening stored caches - Comparison method violates its general contract

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12803]
### Describe your problem!

User on support mail reports, that c:geo crashes as soon as he wants to open his list of stored caches.

```
02-23 20:38:43.009 17204 17204 E AndroidRuntime: FATAL EXCEPTION: main
02-23 20:38:43.009 17204 17204 E AndroidRuntime: Process: cgeo.geocaching, PID: 17204
02-23 20:38:43.009 17204 17204 E AndroidRuntime: java.lang.IllegalArgumentException: Comparison method violates its general contract!
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.TimSort.mergeLo(TimSort.java:777)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.TimSort.mergeAt(TimSort.java:514)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.TimSort.mergeCollapse(TimSort.java:441)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.TimSort.sort(TimSort.java:245)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.Arrays.sort(Arrays.java:1492)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.ArrayList.sort(ArrayList.java:1470)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.util.Collections.sort(Collections.java:206)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at cgeo.geocaching.sorting.AbstractCacheComparator.sort(AbstractCacheComparator.java:82)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at cgeo.geocaching.ui.CacheListAdapter.forceSort(CacheListAdapter.java:340)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at cgeo.geocaching.ui.CacheListAdapter.setList(CacheListAdapter.java:140)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at cgeo.geocaching.CacheListActivity.onLoadFinished(CacheListActivity.java:2095)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at cgeo.geocaching.CacheListActivity.onLoadFinished(CacheListActivity.java:145)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.app.LoaderManagerImpl$LoaderObserver.onChanged(LoaderManagerImpl.java:250)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.lifecycle.LiveData.setValue(LiveData.java:309)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.app.LoaderManagerImpl$LoaderInfo.setValue(LoaderManagerImpl.java:189)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.app.LoaderManagerImpl$LoaderInfo.onLoadComplete(LoaderManagerImpl.java:174)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.content.Loader.deliverResult(Loader.java:132)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.content.AsyncTaskLoader.dispatchOnLoadComplete(AsyncTaskLoader.java:258)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.content.AsyncTaskLoader$LoadTask.onPostExecute(AsyncTaskLoader.java:83)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.content.ModernAsyncTask.finish(ModernAsyncTask.java:490)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at androidx.loader.content.ModernAsyncTask$InternalHandler.handleMessage(ModernAsyncTask.java:507)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8550)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
02-23 20:38:43.009 17204 17204 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
02-23 20:38:43.010 17204 17204 W cgeo    : [main] UncaughtException
02-23 20:38:43.010 17204 17204 W cgeo    : java.lang.IllegalArgumentException: Comparison method violates its general contract!
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.TimSort.mergeLo(TimSort.java:777)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.TimSort.mergeAt(TimSort.java:514)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.TimSort.mergeCollapse(TimSort.java:441)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.TimSort.sort(TimSort.java:245)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.Arrays.sort(Arrays.java:1492)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.ArrayList.sort(ArrayList.java:1470)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.util.Collections.sort(Collections.java:206)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at cgeo.geocaching.sorting.AbstractCacheComparator.sort(AbstractCacheComparator.java:82)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at cgeo.geocaching.ui.CacheListAdapter.forceSort(CacheListAdapter.java:340)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at cgeo.geocaching.ui.CacheListAdapter.setList(CacheListAdapter.java:140)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at cgeo.geocaching.CacheListActivity.onLoadFinished(CacheListActivity.java:2095)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at cgeo.geocaching.CacheListActivity.onLoadFinished(CacheListActivity.java:145)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.app.LoaderManagerImpl$LoaderObserver.onChanged(LoaderManagerImpl.java:250)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.lifecycle.LiveData.setValue(LiveData.java:309)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.app.LoaderManagerImpl$LoaderInfo.setValue(LoaderManagerImpl.java:189)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.app.LoaderManagerImpl$LoaderInfo.onLoadComplete(LoaderManagerImpl.java:174)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.content.Loader.deliverResult(Loader.java:132)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.content.AsyncTaskLoader.dispatchOnLoadComplete(AsyncTaskLoader.java:258)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.content.AsyncTaskLoader$LoadTask.onPostExecute(AsyncTaskLoader.java:83)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.content.ModernAsyncTask.finish(ModernAsyncTask.java:490)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at androidx.loader.content.ModernAsyncTask$InternalHandler.handleMessage(ModernAsyncTask.java:507)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at android.os.Handler.dispatchMessage(Handler.java:106)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at android.os.Looper.loop(Looper.java:246)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at android.app.ActivityThread.main(ActivityThread.java:8550)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at java.lang.reflect.Method.invoke(Native Method)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
02-23 20:38:43.010 17204 17204 W cgeo    : 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

### How to reproduce?

Unclear

### Actual result after these steps?

Crash

### Expected result after these steps?

Normal operation

### Reproducible

Unclear

### c:geo Version

2022.02.16

### System information

```text
## System information

c:geo version: 2022.02.16

Device:
-------
- Device type: SM-A505FN (a50eea, samsung)
- Available processors: 8
- Android version: 11
- Android build: RP1A.200720.012.A505FNXXS9CVA3
- Screen resolution: 1080x2131px (411x811dp)
- Pixel density: 2.625
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: enabled - 21.48.15 (150400-414534850)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Settings: v9, Count:146
- Set language: fi_FI (system default)
- System date format: d.M.y
- Time zone: GMT+02:00
- Debug mode active: no
- Last backup: never
- Routing mode: Walk
- Live map mode: true
- OSM multi-threading: false / threads: 1
- Map: Google: Satelliitti
  - Id: cgeo.geocaching.maps.google.v2.GoogleMapProvider$GoogleSatelliteSource
  - Atts: none
  - Theme: none

Filters:
-------
- Hide waypoints: original
- LIVE: Tila:Löydetty=Ei,Rajaa pois arkistoidut ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status:found_no:exclude_archived))
- OFFLINE: Ei mitään ([:inconclusive=false:advanced=false])

Stored Filters (#0):

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Sisäänkirjautuminen onnistui) / PREMIUM
   Geocaching.com Adventure Lab
- Geocaching.com date format: d.M.yyyy
- Routing: external / BRouter installed: false
- Installed c:geo plugins:  none

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (89,6 GB free) v3 internal isDir(10 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (89,6 GB free) v3 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (89,6 GB free) v3 external non-removable isDir(1588 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (89,6 GB free) v3 internal isDir(0 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=6, dirs:>=8, totalFileSize:>=5,0 MB, free space: 89,6 GB, files on device: 7176000)
  - OFFLINE_MAPS: /cgeo/maps (Default)[/cgeo/maps[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps, Av:true, files:1, dirs:1, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - OFFLINE_MAP_THEMES: /cgeo/maps/_themes (Default)[/cgeo/maps/_themes[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps/_themes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps%2F_themes, Av:true, files:1, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:5, dirs:0, totalFileSize:5,0 MB, free space: 89,6 GB, files on device: 7176000)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:1, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:9, dirs:1, totalFileSize:81,2 KB, free space: 89,6 GB, files on device: 7176000)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: 7176000)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 89,6 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- UNUSED_TRACK: null
- Persisted Uri Permissions: #1
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (24. huhtik. 2021 klo 10.09):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v98, Size:42,4 MB) on system internal storage

--- End of system information ---
```

### Additional Information

Support ticket 105500

**Comment by eddiemuc** [comment]
Is this reproducible? Would be a great help?

Is it known which sorting method was used?

**Comment by Lineflyer** [comment]
I have to ask the user for that.
Lets hope he remembers as he can no longer open the stored list.

Isn't sorting now bound to each list?
Maybe it works if he long presses "stored" and selects another list. But then we still don't see the problematic sorting.

Will ask anyway.

**Comment by eddiemuc** [comment]
> Isn't sorting now bound to each list?

Very good point! The user may open his settings and filter for "sort". He wil find something like the following:

![image](https://user-images.githubusercontent.com/6909759/155513771-70e612fb-d634-4ff6-9982-47c794d26d0b.png)

From this we can see hopefully the currently used sort methods

**Comment by Lineflyer** [comment]
The user confirms he can access other lists by long tap on lists. The faulty list is the default list.
I asked him for the screenshot above from his system and provided step-by-step for that.

After we have that information:
What needs to be set into the fault-causing field to get him going again?

I just tried all sortings without any crash.
So I guess we might additionally need to lists content as a GPX to reproduce?!

**Comment by eddiemuc** [comment]
The standard list's sort setting is the one with key c`ache_sort_config.OFFLINE-1`. I would like to have the value of this setting from the user.

The user should try to set it simply to another setting, preferably something with a rather "simple" sort algo. I assume `NAME-true` should work stable (will sort by name ascending)

The exception might have been triggered by specific values for one or more caches in the user's standard list. Once I know the comparator in question I will analyze it more deeply to see which cache values could cause the inconsistency. But it is quite impossible to analze ALL comparators in this depth.

**Comment by Lineflyer** [comment]
From the type of error: 
Is that kind of a type mismatch (e.g. comparing a string when a number is expected)?

And just for my knowledge:
What does e.g. "NAME-true" or "RATING-false" mean (especially the boolean)?

**Comment by Lineflyer** [comment]
Here we go:
![image](https://user-images.githubusercontent.com/949669/155606379-af54266e-28ba-4b64-b047-7180a0a7ddef.png)

I will ask the user to set the `DIFFICULTY-false` to `AUTO-false` (as this is default) ans send us a GPX export of this default list.

**Comment by eddiemuc** [comment]
So difficulty is the culprits that's unexpected...

> From the type of error: 
> Is that kind of a type mismatch (e.g. comparing a string when a number is expected)?

Sorting routine in java takes as input a "comparator" object which basically implements a method `int compare(Object o1, Object o2)` which has to return an int value less, equal or greater than 0 when `o1` should be considered less, equal or greater than `o2` in the sense of sorting a list of objects.

This method has to follow other guides to be valid as a sort method. The most important thing is that it must be consistent regardless of which objects it is called for. E.g.
* If `compare(o1, o2)` is >0 then `compare (o2, o1)` must be <0
* If `compare (o1,o2)` is >0 and `compare(o2,o3)` is also >0 then `compare (o1,o3)` must also be >0

This is apparentlynot always the case here. First candidate for misfit is usually handling of null values, but we have to look into it.



> And just for my knowledge:
> What does e.g. "NAME-true" or "RATING-false" mean (especially the boolean)?

First part identifies the comparator (there's an enum in code mapping these IDs to comparators, e.g. NAME to name comparator), the second part (boolean) says if sorting should be in ascending or descending order

**Comment by Lineflyer** [comment]
@eddiemuc 

Thanks for explaining.

I do meanwhile have the sqlite backup and can reproduce the problem.
However the DB seems to be valid and the D-ratings looks normal. No idea what dataset is the culprit here.

Can you (or someone else) take a look into the problem. I guess you would need the DB for that.
Will share a download link to cgeo-gev group in a minute.

**Comment by eddiemuc** [comment]
@Lineflyer I can't reproduce the problem with loaded user DB. 

Probiel mmight be SDK-specific too. Which emulator are you using (SDKversion and simulated device)? I am using SDK29 with Pixel 3a

Checking DB everythink looks fine, all caches have a difficulty value. Guessing around:
* I would remove the "catch Exception" in the c:geo sort routine because it might hide bugs in sorting routine as well as creating non-contract situations (which is even commented in code)
* There's a direct compare of difficulty value (float) with the constant 0.0 (double) which I find highly suspicious (one shall never compare float/double values using ==). But can't say if this is the culprit without being able to reproduce problem

![image](https://user-images.githubusercontent.com/6909759/155879411-05a708eb-512e-4586-a944-a54f942d116c.png)

![image](https://user-images.githubusercontent.com/6909759/155879391-09a27dcb-e2c3-469c-b098-c8af064e85a9.png)

**Comment by eddiemuc** [comment]
Addon: I am working on release, is this the correct branch?

**Comment by eddiemuc** [comment]
@Lineflyer I provided PR #12811 which removes both suspicious code parts above. I would be interesting in knowing whether this solves the problem (or at least produces a "better" exception pointing to the real problem - in case the real problem was hidden by the "catch Excception").

**Comment by Lineflyer** [comment]
I used SDK29 Pixel 3a emulator as well.
I restored the users backup as well as the users settings! Might that be a difference?

**Comment by eddiemuc** [comment]
> I used SDK29 Pixel 3a emulator as well.
> I restored the users backup as well as the users settings! Might that be a difference?

Who knows...

I just exchanged the db on device and didn't use the backuo/restore mechanism, maybe this makes a difference as well.

**Comment by Lineflyer** [comment]
Just tested your PR 12811. With that PR the crash is gone!
Building again and testing the same state on current `release` it crashes again.

Would still be interesting to find the root cause...and why you can't reproduce. Related to some setting?

**Comment by Lineflyer** [comment]
BTW:
Difference in setting could be the list load limit. It was set to "unlimited" on the users device. If I reduce it the list loads first and crashes only after loading more caches onto the list (probably as soon as the problematic cache is loaded).
@eddiemuc Is that settings different on your emulator?

**Comment by eddiemuc** [comment]
> Just tested your PR 12811. With that PR the crash is gone!
> Building again and testing the same state on current `release` it crashes again.
> 
> Would still be interesting to find the root cause...and why you can't reproduce. Related to some setting? 

With the PR applied you might try to sort after "terrain". There I did not remove the suspicious comparison with 0.0. if sorting by terrain crashes but sorting by difficulty does not, then this is the root culprit. Also, in case terrain crashes I would be interested in the exact exception (if it is a different one than before which might be the case because I removed the catch exception at the other place)

**Comment by eddiemuc** [comment]
> BTW:
> Difference in setting could be the list load limit. It was set to "unlimited" on the users device. If I reduce it the list loads first and crashes only after loading more caches onto the list (probably as soon as the problematic cache is loaded).
> @eddiemuc Is that settings different on your emulator?

Unlikely that this is the root cause. Inam out in the field unfortunately and can't check

**Comment by Lineflyer** [comment]
> With the PR applied you might try to sort after "terrain". There I did not remove the suspicious comparison with 0.0. if sorting by terrain crashes but sorting by difficulty does not, then this is the root culprit. Also, in case terrain crashes I would be interested in the exact exception (if it is a different one than before which might be the case because I removed the catch exception at the other place)

Sorting by terrain (tested both sort orders) is working without your PR. The crash is limited to difficulty.
To be complete I tested terrain sorting with your PR and it still works.

**Comment by Lineflyer** [comment]
> Unlikely that this is the root cause. Inam out in the field unfortunately and can't check

No, not the root cause but the cause for you failing to be able to reproduce (as not all caches from that list are loaded, thus the compare might work until you loaded all of them).

**Comment by eddiemuc** [comment]
> > Unlikely that this is the root cause. Inam out in the field unfortunately and can't check
> 
> No, not the root cause but the cause for you failing to be able to reproduce (as not all caches from that list are loaded, thus the compare might work until you loaded all of them).

You're right, that was it. I can reproduce now and will try to dig in deeper

**Comment by eddiemuc** [comment]
Now I got it.

Problem was the fallback comparison: if camCompare returns false (which is the case for e.g. difficulty or terrain being 0.0, which is the case for e.g. all Adventure Labs) then comparsion falls back to a comparison of the geocodes.

This comparison in turn uses `GCConstants.gccodeToGCId(cache1.getGeocode())` to calculate a gccode-pseuso-id and substracts them from each other.

Problem was: the GCId returned is a `long`, but the substract was converted to `int`. This conversion could change the signum of the result arbitrarily depending on the combination of Adventure Lans compared -> resulting in inconsistent comparsion results whenever this overflow randomly changed the signum.

Fixed with (revised) PR #12811
